[Templates] Add template "call-for-model" markdown and "call-for-big-bird" markdown (#9921)
huggingface/transformers: templates/adding_a_new_model/ADD_NEW_MODEL_PROPOSAL_TEMPLATE.md

**TEMPLATE**
=====================================

> *search & replace the following keywords, e.g.:*

Reviewer: Sounds like a job for cookiecutter ;-)

`:%s/\[name of model\]/brand_new_bert/g`

-[lowercase name of model]  # e.g. brand_new_bert

-[camelcase name of model]  # e.g. BrandNewBert

-[name of mentor]  # e.g. Peter

-[link to original repo]

-[start date]

-[end date]

How to add [camelcase name of model] to 🤗 Transformers?
=====================================

Mentor: [name of mentor]

Begin: [start date]

Estimated End: [end date]

Adding a new model is often difficult and requires an in-depth knowledge

> of the 🤗 Transformers library and ideally also of the model\'s original

Reviewer: ```suggestion
of the 🤗 Transformers library and ideally also of the model's original
```

Reviewer: There are lots of those, I don't get why.

reasonable level.

A good first starting point to better understand the library is to read
the [documentation of our philosophy](https://huggingface.co/transformers/philosophy.html).
As a result of our way of working, there are some choices that we try to apply to all models:

-   Composition is generally favored over-abstraction
-   Duplicating code is not always bad if it strongly improves the
    readability or accessibility of a model
-   Model files are as self-contained as possible so that when you read
    the code of a specific model, you ideally only have to look into the
    respective `modeling_....py` file.

In our opinion, the library\'s code is not just a means to provide a
product, *e.g.* the ability to use BERT for inference, but also as the
very product that we want to improve. Hence, when adding a model, the
user is not only the person that will use your model, but also everybody
that will read, try to understand, and possibly tweak your code.

With this in mind, let\'s go a bit deeper into the general library
design.

### Overview of models

To successfully add a model, it is important to understand the
interaction between your model and its config,

> `PreTrainedModel`, and `.PretrainedConfig`. For

Reviewer: ```suggestion
`PreTrainedModel`, and `PretrainedConfig`. For
```

exemplary purposes, we will call the model to be added to 🤗 Transformers

> `BrandNewBert`.

Reviewer: We have the name of the model here, why not use it?

model are
`PreTrainedModel.from_pretrained` and `PreTrainedModel.save_pretrained`, which are 
used for serialization and deserialization. All
of the other important functionalities, such as
`BrandNewBertModel.forward` should be
completely defined in the new `modeling_brand_new_bert.py` script. Next,
we want to make sure that a model with a specific head layer, such as
`BrandNewBertForMaskedLM` does not inherit
from `BrandNewBertModel`, but rather uses
`BrandNewBertModel` as a component that
can be called in its forward pass to keep the level of abstraction low.
Every new model requires a configuration class, called
`BrandNewBertConfig`. This configuration
is always stored as an attribute in
`PreTrainedModel`, and
thus can be accessed via the `config` attribute for all classes
inheriting from `BrandNewBertPreTrainedModel`

```python
model = BrandNewBertModel.from_pretrained("brandy/brand_new_bert")
model.config  # model has access to its config
```

Similar to the model, the configuration inherits basic serialization and
deserialization functionalities from
`PretrainedConfig`. Note
that the configuration and the model are always serialized into two

> different formats - the model to a `pytorch\_model.bin` file

Reviewer: ```suggestion
different formats - the model to a `pytorch_model.bin` file
```
Same as for the \'.

understand the theoretical aspects of the model *on-the-job*, then it is
totally fine to directly dive into the `[camelcase name of model]`\'s code-base. This
option might suit you better, if your engineering skills are better than
your theoretical skill, if you have trouble understanding
`[camelcase name of model]`\'s paper, or if you just enjoy programming much more than
reading scientific papers.

### 1. (Optional) Theoretical aspects of [camelcase name of model]

You should take some time to read *[camelcase name of model]\'s* paper, if such
descriptive work exists. There might be large sections of the paper that
are difficult to understand. If this is the case, this is fine - don\'t
worry! The goal is not to get a deep theoretical understanding of the
paper, but to extract the necessary information required to effectively
re-implement the model in 🤗 Transformers. That being said, you don\'t
have to spend too much time on the theoretical aspects, but rather focus
on the practical ones, namely:

-   What type of model is *[camelcase name of model]*? BERT-like encoder-only
    model? GPT2-like decoder-only model? BART-like encoder-decoder

>     model? Look at the `model_summary`{.interpreted-text role="doc"} if

Reviewer: What does `{.interpreted-text role="doc"}` mean?

You should have understood the following aspects of [camelcase name of model] by now:

- [characteristic 1 of [camelcase name of model]]
- [characteristic 2 of [camelcase name of model]]
- ...

If any of the mentioned aspects above are **not** clear to you, now is a great time to talk to [name of mentor].

### 2. Next prepare your environment

1.  Fork the [repository](https://github.com/huggingface/transformers)
    by clicking on the 'Fork\' button on the repository\'s page. This
    creates a copy of the code under your GitHub user account.

2.  Clone your `transformers` fork to your local disk, and add the base
    repository as a remote:

    ```bash
    git clone https://github.com/[your Github handle]/transformers.git
    cd transformers
    git remote add upstream https://github.com/huggingface/transformers.git
    ```

3.  Set up a development environment, for instance by running the
    following command:

    ```bash
    python -m venv .env
    source .env/bin/activate
    pip install -e ".[dev]"
    ```

and return to the parent directory

```bash
cd ..
```

4.  We recommend adding the PyTorch version of *[camelcase name of model]* to

>     Transformers. To install PyTorch, please follow the instructions on
>     <https://pytorch.org/get-started/locally/>.

Reviewer: ```suggestion
    Transformers. To install PyTorch, please follow the instructions [here](https://pytorch.org/get-started/locally/).
```

user-friendly, and beautiful** as possible. This is the number-one
motivation to re-implement models into 🤗 Transformers - trying to make
complex new NLP technology accessible to **everybody**.

You should start thereby by diving into the [original repository]([link to original repo]).

Successfully running the official pretrained model in the original
repository is often **the most difficult** step. From our experience, it
is very important to spend some time getting familiar with the original
code-base. You need to figure out the following:

-   Where to find the pretrained weights?
-   How to load the pretrained weights into the corresponding model?
-   How to run the tokenizer independently from the model?
-   Trace one forward pass so that you know which classes and functions
    are required for a simple forward pass. Usually, you only have to
    reimplement those functions.
-   Be able to locate the important components of the model: Where is
    the model\'s class? Are there model sub-classes, *e.g.*
    EncoderModel, DecoderModel? Where is the self-attention layer? Are
    there multiple different attention layers, *e.g.* *self-attention*,

>     *cross-attention*\...?

Reviewer: ```suggestion
    *cross-attention*...?
```

the original model

-   [Jupyter notebooks](https://jupyter.org/) / [google colab](https://colab.research.google.com/notebooks/intro.ipynb)
-   Local python scripts.

Jupyter notebooks have the advantage that they allow for cell-by-cell
execution which can be helpful to better split logical components from
one another and to have faster debugging cycles as intermediate results
can be stored. Also, notebooks are often easier to share with other
contributors, which might be very helpful if you want to ask the Hugging
Face team for help. If you are familiar with Jupiter notebooks, we
strongly recommend you to work with them.

The obvious disadvantage of Jupyther notebooks is that if you are not
used to working with them you will have to spend some time adjusting to
the new programming environment and that you might not be able to use
your known debugging tools anymore, like `ipdb`.

For each code-base, a good first step is always to load a **small**
pretrained checkpoint and to be able to reproduce a single forward pass
using a dummy integer vector of input IDs as an input. Such a script
could look like this (in pseudocode):

```python

> model = [camelcase name of model]Model.load_pretrained_checkpoint(/path/to/checkpoint/)

Reviewer: ```suggestion
model = [camelcase name of model]Model.load_pretrained_checkpoint("/path/to/checkpoint/")
```

 [-0.4417, -0.5920,  0.3450,  ..., -0.3062,  0.6182,  0.7132],
 [-0.5009, -0.7122,  0.4548,  ..., -0.3662,  0.6091,  0.7648],
 ...,
 [-0.5613, -0.6332,  0.4324,  ..., -0.3792,  0.7372,  0.9288],
 [-0.5416, -0.6345,  0.4180,  ..., -0.3564,  0.6992,  0.9191],
 [-0.5334, -0.6403,  0.4271,  ..., -0.3339,  0.6533,  0.8694]]],
```

We expect that every model added to 🤗 Transformers passes a couple of
integration tests, meaning that the original model and the reimplemented
version in 🤗 Transformers have to give the exact same output up to a
precision of 0.001! Since it is normal that the exact same model written
in different libraries can give a slightly different output depending on
the library framework, we accept an error tolerance of 1e-3 (0.001). It
is not enough if the model gives nearly the same output, they have to be
the almost identical. Therefore, you will certainly compare the
intermediate outputs of the 🤗 Transformers version multiple times
against the intermediate outputs of the original implementation of
*[camelcase name of model]* in which case an **efficient** debugging environment
of the original repository is absolutely important. Here is some advice

> is to make your debugging environment as efficient as possible.

Reviewer 2: ```suggestion
to make your debugging environment as efficient as possible.
```

    Jax? Then make sure that the model is **not jitted** when running
    the forward pass, *e.g.* check-out [this
    link](https://github.com/google/jax/issues/196).
-   Use the smallest pretrained checkpoint you can find. The smaller the
    checkpoint, the faster your debug cycle becomes. It is not efficient
    if your pretrained model is so big that your forward pass takes more
    than 10 seconds. In case only very large checkpoints are available,
    it might make more sense to create a dummy model in the new
    environment with randomly initialized weights and save those weights
    for comparison with the 🤗 Transformers version of your model
-   Make sure you are using the easiest way of calling a forward pass in
    the original repository. Ideally, you want to find the function in
    the original repository that **only** calls a single forward pass,
    *i.e.* that is often called `predict`, `evaluate`, `forward` or
    `__call__`. You don\'t want to debug a function that calls `forward`
    multiple times, *e.g.* to generate text, like
    `autoregressive_sample`, `generate`.
-   Try to separate the tokenization from the model\'s

>     [forward]{.title-ref} pass. If the original repository shows

Reviewer 2: Here too, weird formatting

    examples where you have to input a string, then try to find out
    where in the forward call the string input is changed to input ids
    and start from this point. This might mean that you have to possibly
    write a small script yourself or change the original code so that
    you can directly input the ids instead of an input string.
-   Make sure that the model in your debugging setup is **not** in
    training mode, which often causes the model to yield random outputs
    due to multiple dropout layers in the model. Make sure that the
    forward pass in your debugging environment is **deterministic** so
    that the dropout layers are not used. Or use

>     [transformers.file\_utils.set\_seed]{.title-ref} if the old and new

Reviewer 2: weird formatting

code is fully correct or clean. Rather, it is advised to add a first
*unclean*, copy-pasted version of the original code to
`src/transformers/models/[lowercase name of model]/modeling_[lowercase name of model].py`
until you feel like all the necessary code is added. From our
experience, it is much more efficient to quickly add a first version of
the required code and improve/correct the code iteratively with the
conversion script as described in the next section. The only thing that
has to work at this point is that you can instantiate the 🤗 Transformers
implementation of *[camelcase name of model]*, *i.e.* the following command
should work:

```python
from transformers import [camelcase name of model]Model, [camelcase name of model]Config
model = [camelcase name of model]Model([camelcase name of model]Config())
```

The above command will create a model according to the default
parameters as defined in `[camelcase name of model]Config()` with random weights,
thus making sure that the `init()` methods of all components works.

[Here the mentor should add very specific information on what exactly has to be changed for this model]
[...]

> [...]

Reviewer: Maybe make it more easy to search with "TODO FILL" and make a mention at the beginning to search for this pattern?

either you used incorrect parameters in `[camelcase name of model]Config()`, have a
wrong architecture in the 🤗 Transformers implementation, you have a bug
in the `init()` functions of one of the components of the 🤗 Transformers
implementation or you need to transpose one of the checkpoint weights.

This step should be iterated with the previous step until all weights of
the checkpoint are correctly loaded in the Transformers model. Having
correctly loaded the checkpoint into the 🤗 Transformers implementation,
you can then save the model under a folder of your choice
`/path/to/converted/checkpoint/folder` that should then contain both a
`pytorch_model.bin` file and a `config.json` file:

```python
model.save_pretrained("/path/to/converted/checkpoint/folder")
```

[Here the mentor should add very specific information on what exactly has to be done for the conversion of this model]
[...]
[...]

**7. Implement the forward pass**

Having managed to correctly load the pretrained weights into the 🤗
Transformers implementation, you should now make sure that the forward
pass is correctly implemented. In [Get familiar with the original

> repository](#run-a-pretrained-checkpoint-using-the-original-repository),

Reviewer: How this will Markdown be rendered? I'm not sure if the link will be easy to see.

Author: fixed it :-)

you have already created a script that runs a forward pass of the model
using the original repository. Now you should write an analogous script
using the 🤗 Transformers implementation instead of the original one. It
should look as follows:

[Here the model name might have to be adapted, *e.g.* maybe [lowercase name of model]ForConditionalGeneration instead of [lowercase name of model]Model]

```python
model = [camelcase name of model]Model.from_pretrained(/path/to/converted/checkpoint/folder)
input_ids = [0, 4, 4, 3, 2, 4, 1, 7, 19]
output = model(input_ids).last_hidden_states
```

It is very likely that the 🤗 Transformers implementation and the
original model implementation don\'t give the exact same output the very
first time or that the forward pass throws an error. Don\'t be
disappointed - it\'s expected! First, you should make sure that the
forward pass doesn\'t throw any errors. It often happens that the wrong

> dimensions are used leading to a [Dimensionality mismatch]{.title-ref}

Reviewer: Other weird thing

instead of `torch.float32`. Don\'t hesitate to ask [name of mentor]
for help, if you don\'t manage to solve certain errors.

The final part to make sure the 🤗 Transformers implementation works
correctly is to ensure that the outputs are equivalent to a precision of
`1e-3`. First, you should ensure that the output shapes are identical,
*i.e.* `outputs.shape` should yield the same value for the script of the
🤗 Transformers implementation and the original implementation. Next, you
should make sure that the output values are identical as well. This one
of the most difficult parts of adding a new model. Common mistakes why
the outputs are not identical are:

-   Some layers were not added, *i.e.* an [activation]{.title-ref} layer
    was not added, or the residual connection was forgotten
-   The word embedding matrix was not tied
-   The wrong positional embeddings are used because the original
    implementation uses on offset
-   Dropout is applied during the forward pass. To fix this make sure
    [model.training is False]{.title-ref} and that no dropout layer is
    falsely activated during the forward pass, *i.e.* pass
    [self.training]{.title-ref} to [PyTorch\'s functional

>     dropout](https://pytorch.org/docs/stable/nn.functional.html?highlight=dropout#torch.nn.functional.dropout)

Reviewer: Same here, lots of weird formatting with this {.title-ref}. (Searching for those would help to make sure we didn't forget any during the review.)

implement the model to 🤗 Transformers. A template of those model tests
is already added by the Cookiecutter, called
`[camelcase name of model]ModelIntegrationTests` and only has to be filled out by
you. To ensure that those tests are passing, run

```python
RUN_SLOW=1 pytest -sv tests/test_modeling_[lowercase name of model].py::[camelcase name of model]ModelIntegrationTests
```

**Note:** In case you are using Windows, you should replace `RUN_SLOW=1` with `SET RUN_SLOW=1`

Second, all features that are special to *[camelcase name of model]* should be
tested additionally in a separate test under
`[camelcase name of model]ModelTester`/`[camelcase name of model]ModelTest`. This part is often
forgotten but is extremely useful in two ways:

-   It helps to transfer the knowledge you have acquired during the
    model addition to the community by showing how the special features
    of *[camelcase name of model]* should work.
-   Future contributors can quickly test changes to the model by running
    those special tests.

[Here the mentor should add very specific information on what special features of the model should be tested additionally]
[...]

> [...]

Reviewer: Same remark as above -> TODO FILL

**9. Implement the tokenizer**

Next, we should add the tokenizer of *[camelcase name of model]*. Usually, the
tokenizer is equivalent or very similar to an already existing tokenizer
of 🤗 Transformers.

[Here the mentor should add a comment whether a new tokenizer is required or if this is not the case which existing tokenizer closest resembles 
 [camelcase name of model]'s tokenizer and how the tokenizer should be implemented]
 [...]

>  [...]

Reviewer: Same remark as above -> TODO FILL

It is very important to find/extract the original tokenizer file and to
manage to load this file into the 🤗 Transformers\' implementation of the
tokenizer.

For [camelcase name of model], the tokenizer files can be found here:

> - [To be filled out by mentor]

Reviewer: Same remark as above -> TODO FILL

and having implemented the  🤗Transformers' version of the tokenizer can be loaded as follows:

> [To be filled out by mentor]

Reviewer: Same remark as above -> TODO FILL
